ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,135 @@
+    
+    free(s_entries);
+    free(emptyArray);
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsStartingIndexWhenNull {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(cacheIndex == startingIndex);
+    
+    free(s_entries);
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsFalseWhenCacheFull {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    memset(s_entries, 1, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertFalse(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    
+    free(s_entries);
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_WrapsToBeginningWhenResolvingCollisions {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    memset(s_entries, 1, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    uintptr_t emptyIndex = 2;
+    s_entries[emptyIndex].vnode = nullptr;
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(emptyIndex == cacheIndex);
+    
+    free(s_entries);
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsLastIndexWhenEmptyAndResolvingCollisions {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    memset(s_entries, 1, s_entriesCapacity * sizeof(VnodeCacheEntry));
+    uintptr_t emptyIndex = s_entriesCapacity - 1;
+    s_entries[emptyIndex].vnode = nullptr;
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(emptyIndex == cacheIndex);
+    
+    free(s_entries);
+}
+
+- (void)testTryInsertOrUpdateEntry_ExclusiveLocked_ReturnsFalseWhenFull {
+    s_entriesCapacity = 100;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    memset(s_entries, 1, s_entriesCapacity * sizeof(VnodeCacheEntry));
+
+    uintptr_t indexFromHash = 5;
+    vnode_t testVnode = &TestVnode;
+    uint32_t testVnodeVid = 7;
+
+    XCTAssertFalse(
+        TryInsertOrUpdateEntry_ExclusiveLocked(
+            testVnode,
+            indexFromHash,
+            testVnodeVid,
+            true, // invalidateEntry
+            TestRootHandle));
+
+    free(s_entries);
+}
```

Comment:
If we aren't able to simplify invalidateEntry, it would be great to have cases to validate it.